go/p2p/peermgmt/registry.go

```diff
@@ -0,0 +1,277 @@
+package peermgmt
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-libp2p/core"
+	"github.com/libp2p/go-libp2p/core/peer"
+	manet "github.com/multiformats/go-multiaddr/net"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/common/node"
+	consensus "github.com/oasisprotocol/oasis-core/go/consensus/api"
+	"github.com/oasisprotocol/oasis-core/go/p2p/api"
+	"github.com/oasisprotocol/oasis-core/go/p2p/rpc"
+	"github.com/oasisprotocol/oasis-core/go/p2p/txsync"
+	kmp2p "github.com/oasisprotocol/oasis-core/go/worker/keymanager/p2p"
+	"github.com/oasisprotocol/oasis-core/go/worker/storage/p2p/pub"
+	storageSync "github.com/oasisprotocol/oasis-core/go/worker/storage/p2p/sync"
+)
+
+type peerRegistry struct {
+	logger *logging.Logger
+
+	mu sync.Mutex
+
+	consensus    consensus.Backend
+	chainContext string
+
+	peers map[core.PeerID]*peer.AddrInfo
+
+	protocolPeers map[core.ProtocolID]map[core.PeerID]*peer.AddrInfo
+	topicPeers    map[string]map[core.PeerID]*peer.AddrInfo
+
+	initCh   chan struct{}
+	initOnce sync.Once
+}
+
+func newPeerRegistry(c consensus.Backend, chainContext string) *peerRegistry {
+	l := logging.GetLogger("p2p/peer-manager/registry")
+
+	return &peerRegistry{
+		logger:        l,
+		consensus:     c,
+		chainContext:  chainContext,
+		peers:         make(map[core.PeerID]*peer.AddrInfo),
+		protocolPeers: make(map[core.ProtocolID]map[core.PeerID]*peer.AddrInfo),
+		topicPeers:    make(map[string]map[core.PeerID]*peer.AddrInfo),
+		initCh:        make(chan struct{}),
+	}
+}
+
+// Implements api.PeerRegistry.
+func (r *peerRegistry) Initialized() <-chan struct{} {
+	return r.initCh
+}
+
+// Implements api.PeerRegistry.
+func (r *peerRegistry) NumPeers() int {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return len(r.peers)
+}
+
+// Implements api.RegistryManager.
+func (r *peerRegistry) ProtocolPeers(p core.ProtocolID) []*peer.AddrInfo {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	pp := r.protocolPeers[p]
+	peers := make([]*peer.AddrInfo, 0, len(pp))
+	for _, peer := range pp {
+		peers = append(peers, peer)
+	}
+
+	return peers
+}
+
+// Implements api.RegistryManager.
+func (r *peerRegistry) TopicPeers(topic string) []*peer.AddrInfo {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	tp := r.topicPeers[topic]
+	peers := make([]*peer.AddrInfo, 0, len(tp))
+	for _, peer := range tp {
+		peers = append(peers, peer)
```

Comment:
I think this should append a copy of `AddrInfo` and not the pointer directly.  Otherwise, `TopicPeers` is not thread-safe and the caller can also modify the value in the original `topicPeers`.

Reply:
Copy won't work, as one can always mess with the addrs slice inside `AddrInfo`. As I don't want to copy these things, I decided to lowercase the method and keep it private. The package should treat it as read-only, as it already does.